### PR TITLE
fix/13945_oplog_tailing_grouped_operations

### PR DESCRIPTION
**Pull request by turoar23**
As discussed here https://github.com/meteor/meteor/issues/13945 , the current oplog tailing filter does not correctly handle grouped operations. We need to update the query accordingly.

We can receive documents like the following:
```
{
  "txnNumber": 1,
  "op": "c",
  "ns": "admin.$cmd",
  applyOps: [
    { op: "i", ns: "db1.collection1", o: { _id: 1, name: "Alice" } },
    { op: "u", ns: "db2.collection2", o: { $set: { age: 30 } }, o2: { _id: 1 } },
    { op: "d", ns: "db3.collection3", o2: { _id: 1 } }
  ]
```
The idea is to apply the filtering first in the listener.
For mixed operations (e.g. transactions where multiple collections are grouped inside applyOps), we should **not** filter out the document if at least one of the collections should be processed. Later, during document analysis, the filtering will be applied more precisely, so the unwanted operations won’t be sent through DDP.

### CODE REVIEW

File: packages/mongo/oplog_tailing.ts [hunk 1]
```diff
@@ -104,40 +127,55 @@ export class OplogHandle {
...
-          $regex: nsRegex,
-          $nin: this._oplogOptions.excludeCollections.map(
-            (collName: string) => `${this._dbName}.${collName}`
-          ),
-        },
+        $or: [
+          { ns: excludeNs },
+          {
+            ns: /^admin\.\$cmd/,
+            'o.applyOps': { $elemMatch: { ns: { $nin: excludeNs.$nin } } },
```
Comment (dupontbertrand) [inline packages/mongo/oplog_tailing.ts]: For the exclude case, I think it’s safer to keep $regex inside the elemMatch, otherwise sub-ops from other databases or unrelated namespaces might leak through ? (I haven't tested)
```
"o.applyOps": { 
     $elemMatch: { ns: { $regex: nsRegex, $nin: excludeNsList } }
} 
```

**Comment by Twisterking** [comment]
Nice job @turoar23 ! 🥳 
I would definitely suggest to extend the tests for this case! Maybe you can create a test that writes 1k documents with a bulk insert into 2 collections and check if the oplog entries are correctly ignored with the specific settings! Having tests for this setup seems crucial to me as Mongo 8 seems to use this more and more (as you outlined). Well done! ✌️

File: packages/mongo/oplog_tailing.ts (hunk 1 above)
Comment (turoar23) [reply packages/mongo/oplog_tailing.ts]: You are right, other operations will get through and in the others cases we are filtering them. I will change it! Good catch

**Comment by turoar23** [comment]
> Nice job @turoar23 ! 🥳 I would definitely suggest to extend the tests for this case! Maybe you can create a test that writes 1k documents with a bulk insert into 2 collections and check if the oplog entries are correctly ignored with the specific settings! Having tests for this setup seems crucial to me as Mongo 8 seems to use this more and more (as you outlined). Well done! ✌️

Thank you 😄

That’s a great idea — I’ll try to add tests for the different cases we discussed, like transactions, massive inserts, and so on.

**Comment by dupontbertrand** [comment]
@turoar23 I think we should also add tests for operations like rename, drop, and dropDatabase.
My concern is that with the current filters, these admin.$cmd ops may never reach handleDoc, since they are not applyOps. Even though the logic to handle them exists at the bottom of the file, they might get excluded by the include/exclude selectors here:

```
oplogCriteria.push({
  $or: [
    { ns: excludeNs },
    {
      ns: /^admin\.\$cmd/,
      'o.applyOps': { $elemMatch: { ns: excludeNs } },
    },
  ],
});
```


and

```
oplogCriteria.push({
  $or: [
    { ns: /^admin\.\$cmd/, 'o.applyOps.ns': includeNs },
    { ns: includeNs },
  ],
});
```


(Also, maybe worth formatting them so both have a consistent style :wink: )

**Comment by turoar23** [comment]
> @turoar23 I think we should also add tests for operations like rename, drop, and dropDatabase. My concern is that with the current filters, these admin.$cmd ops may never reach handleDoc, since they are not applyOps. Even though the logic to handle them exists at the bottom of the file, they might get excluded by the include/exclude selectors here:

@dupontbertrand I will check it, but I see we have this first filter:
```
const oplogCriteria: any = [
      {
        $or: [
          { op: { $in: ["i", "u", "d"] } },
          { op: "c", "o.drop": { $exists: true } },
          { op: "c", "o.dropDatabase": 1 },
          { op: "c", "o.applyOps": { $exists: true } },
        ],
      },
    ];
```

So I understand the new change won't affect rename, drop, and dropDatabase. However, I think it's a good idea to add these tests.

**Comment by turoar23** [comment]
@Twisterking I added tests for massive inserts and transactions. With the current version they fail, but with these new changes they are now passing 😄

**Comment by nachocodoner** [comment]
@turoar23: Thank you for the work on solving these use cases.  🚀

This is an excellent first contribution, with tests included to cover the edge case, making it easier to understand the problem. This fix will surely help many users relying on the pure Meteor OPLOG driver and collection exclusions.

Also great to see comments from @Twisterking and @dupontbertrand helping guide the final solution.

I’m marking this change to be included as a contribution for [Meteor 3.4](https://github.com/meteor/meteor/pull/13915). Currently, Meteor 3.4 is in beta, mainly focused on the [Rspack integration](https://github.com/meteor/meteor/pull/13910). Soon, we’ll release new betas that include more contributions like this one. I’ll let you know here once it’s available for beta testing.

**Review by dupontbertrand — APPROVED** [review]

**Review by nachocodoner — APPROVED** [review]
